llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,443 @@
+  const MachineRegisterInfo *MRI = nullptr;
+  const AIEBaseRegisterInfo *TRI = nullptr;
+  VirtRegMap *VRM = nullptr;
+  LiveRegMatrix *LRM = nullptr;
+  LiveIntervals *LIS = nullptr;
+
+  bool
+  renameMBBPhysRegs(const SmallVector<const MachineBasicBlock *, 4> &LoopBBs);
+  SmallSet<MCPhysReg, 16>
+  getBBForbiddenReplacements(const MachineBasicBlock *MBB);
+
+  bool replaceReg(const Register Reg,
+                  SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  /// Find Free register of the same register Class type and
+  // exclude the forbidden Physical register from the result
+  SmallVector<MCPhysReg, 16>
+  getPriorityReplacementRegs(const Register Reg,
+                             const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs);
+
+  bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
+                      const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                      const SmallVector<MCPhysReg, 16> &UsedVirtRegs);
+
+  bool isIdentityCopy(const MachineInstr &MI);
+
+  // Check if the UnDef is the last define for a Register, or if there is
+  // another Def/Undef that will later complete the Definition of the
+  // Register. If so, it is too early to determine, if the Register is a
+  // candidate for a WAW replacement, or not. This is especially interesting,
+  // if Subregisters are used to hold a constant and this shared
+  // subregister-constant is equal for multiple Registers. A renamig here
+  // would destroy the shared usage of the sub-register and force the
+  // insertion of redundant Instructions.
+  bool isUndefReserved(
+      const MachineOperand &Def, const MachineInstr &MI,
+      const MachineBasicBlock &MBB,
+      std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse);
+
+  bool isNotLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                    const MachineBasicBlock &MBB);
+
+  bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                    SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+  bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                              const MachineInstr *InstrB);
+
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+};
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  bool Modified = renameMBBPhysRegs(LoopBBs);
```

Comment:
Should we collapse as `return renameMBBPhysRegs(LoopBBs);`?

Reply:
yes, this was a remnant of a debug run